presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/RemoveRedundantTopN.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.sql.planner.iterative.rule;
+
+import io.prestosql.matching.Captures;
+import io.prestosql.matching.Pattern;
+import io.prestosql.sql.planner.iterative.Rule;
+import io.prestosql.sql.planner.plan.TopNNode;
+
+import static io.prestosql.sql.planner.optimizations.QueryCardinalityUtil.isAtMost;
+import static io.prestosql.sql.planner.plan.Patterns.TopN.count;
+import static io.prestosql.sql.planner.plan.Patterns.topN;
+
+/**
+ *  Remove TopN node when the subplan is guaranteed to produce fewer rows than the TopN count
+ */
+public class RemoveRedundantTopN
+        implements Rule<TopNNode>
+{
+    private static final Pattern<TopNNode> PATTERN = topN().with(count().matching(count -> count != 0));
+
+    @Override
+    public Pattern<TopNNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(TopNNode node, Captures captures, Context context)
+    {
+        if (isAtMost(node.getSource(), context.getLookup(), node.getCount())) {
+            return Result.ofPlanNode(node.getSource());
```

Comment:
TopN cannot be removed blindly. The ordering matters. You can replace it with a SortNode in this case.

It's only safe to remove if the row count is guaranteed to be 1.